### ISSUE TITLE
Highlightr not working in Obsidian v.15.6 (last update)

### ISSUE THREAD

**Issue by gabrielcarus** [chetachiezikeuzor/Highlightr-Plugin#42]
Hey Chetachi!
Just leaving some issues on the plugins here, so you can take a look later, take your time to deal with your personal problems, no need to hurry!
First of all, congrats on the obsidian plugins, very nice job you did, I can see that you putted much love and effort in every single one of them. Thank you a lot. 

Alright, so now the issues:
- Highlightr stopped working in Obsidian v.15.6, now we are just able to highlight text through the plugin with the hotkeys. 
- An old problem with Highlightr, that I believe you already know, is the space after the marked text.
- A small detail, that would be nice to analyze, is that cMenu is taking a lot to load.
- Snippets plugin didn't stop, but is having some basic bugs in its structure (not making possible to create new CSS Snippets through the plugin and the "hovering" when using the mouse is also bugged).

As I said, take your time alright? 
That's it, best of luck 🫂

**Comment by rossoj85** [comment]
Can confirm. This just stopped working for me about 20 minutes ago.

**Comment by wutsminame** [comment]
confirm

**Comment by roygu22** [comment]
same here

**Comment by toroidalGames** [comment]
Isn't working for me either

**Comment by Xieleilab** [comment]
confirm

**Comment by iMagUdspEllr** [comment]
The shortcuts and the right-click menu do not work.

**Comment by pommes99** [comment]
Same here. Relieved to find this posting and hope it works again!

**Comment by sialaweb** [comment]
Yes confirm, the right-click on highlight don't do anything

**Comment by duanqiaobb** [comment]
confirm here.

**Comment by Metroid-Prime** [comment]
Just found this plugin and not working for me, ugh.

**Comment by daamiian** [comment]
I'm now getting "focus must be in editor"

**Comment by Paddy1058** [comment]
FYI, I found a partial work a round -- by setting hotkeys for the various colors, I can now again add highlights. But setting a hot key to open the highlightr doesn't work -- it just doesn't respond, and doesn't display the choice of colors. Also, to unlighlight, right-clicking on the highlight doesn't bring up the unhighlight menu, but raises a note that "focus must be in editor" (I'm using live preview). Clicking on the highlight puts it in edit mode, then right-clicking brings up unhighlight option, and clicking then unhighlights. Hope this may help in troubleshooting. Really like this plugin when it was working normally.

See my post #41

**Comment by daamiian** [comment]
Another workaround for those who don't want to make new hotkeys:

Highlight the word, open command palette, "Highlightr" - pick a color. 

This takes longer, but I have enough hotkeys as it is, so this works for me as a temporary solution.

**Comment by CharlesCollin91** [comment]
Any news about that issue ?

**Comment by KraXen72** [comment]
it is not working, +1

**Comment by zokizuan** [comment]
it is not working, +1

**Comment by huyz** [comment]
Hmm, single keyboard shortcut to a color works for me as a workaround

**Comment by mikelyncheski** [comment]
To be clear:

(1) Select text
(2) Right click
(3) Choose "Highlight"
(4) Nadda, nothing, zippo

Control P works fine.

**Comment by jasperdj** [comment]
Still broken on obsidian 0.15.9, highlightr 1.2.1.

**Comment by ikuygithub** [comment]
I've just installed it yesterday and was struggling for hours. 
wish I noticed that its no longer available much earlier

**Comment by SimoAbjaou** [comment]
Its not working for me any more with obsidian 15.6

**Comment by KraXen72** [comment]
@chetachiezikeuzor could we please get a fix? thanks

**Comment by remerle** [comment]
For a workaround for the broken context menu (for the Desktop app), go to your `.obsidian\plugins` folder or `Settings -> Community Plugins -> Installed Plugins` and click the folder icon:
![image](https://user-images.githubusercontent.com/1167333/182479386-cddacec5-9cb2-4ccb-abf6-21b2929f32da.png)

Go into the `highlightr-plugin` folder and open `main.js`. Search for `menu.showAtPosition` and replace the following:

```javascript
        menu.showAtPosition({
            x: coords.right + 25,
            y: coords.top + 20,
        });
```

with

```javascript
        const orig=menu.__proto__.sort; // Save the original function
        menu.__proto__.sort = function() {}; // Replace the sort function with a no-op function
        menu.showAtPosition({
            x: coords.right + 25,
            y: coords.top + 20,
        });
       menu.__proto__.sort=orig; // Reset the sort function
```

Basically, the `menu.sort` function relies on menu items which this plugin just ignores. So, when it hits `menu.sort`, the dropdown is rebuilt and the dom (which this plugin manipulates) is reset. This change just makes the `sort` function a no-op.

**Comment by ChriiMoral** [comment]
I'd like to confirm this is happening to me as well. The context menu doesn't work, but commands do.

**Comment by rossoj85** [comment]
> For a workaround for the broken context menu (for the Desktop app), go to your `.obsidian\plugins` folder or `Settings -> Community Plugins -> Installed Plugins` and click the folder icon: ![image](https://user-images.githubusercontent.com/1167333/182479386-cddacec5-9cb2-4ccb-abf6-21b2929f32da.png)
> 
> Go into the `highlightr-plugin` folder and open `main.js`. Search for `menu.showAtPosition` and replace the following:
> 
> ```js
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
> ```
> 
> with
> 
> ```js
>         const orig=menu.__proto__.sort; // Save the original function
>         menu.__proto__.sort = function() {}; // Replace the sort function with a no-op function
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
>        menu.__proto__.sort=orig; // Reset the sort function
> ```
> 
> Basically, the `menu.sort` function relies on menu items which this plugin just ignores. So, when it hits `menu.sort`, the dropdown is rebuilt and the dom (which this plugin manipulates) is reset. This change just makes the `sort` function a no-op.

Tried this, but didn't work. Still getting the same "focus must be in editor error message"

**Comment by dasu88** [comment]
> menu.showAtPosition

According to this method of modification, it works for me, but it needs to be closed once to restart the program
![2022-08-05_103623](https://user-images.githubusercontent.com/23251263/182990033-34e09fc9-c0af-4c45-a976-e0a7c5f87567.jpg)

**Comment by CharlesCollin91** [comment]
Working for me to ! Just got to restart my computer, only restart obsidian didn't work.

**Comment by rossoj85** [comment]
Still no luck on my end...  

Seems to be skipping over that menu.shotAtPosition block and going directly to the error message. Logged it out in the console to see what is going on.  Anyone else on a Mac? Could that have something to do with it?

<img width="671" alt="image" src="https://user-images.githubusercontent.com/25422645/183221194-6c62bd86-331b-4de9-9beb-c1650843346a.png">



<img width="676" alt="image" src="https://user-images.githubusercontent.com/25422645/183221089-6a5e1a51-d021-4778-bcc9-fad62fc5d4f3.png">

**Comment by ObsidianTTRPGProject** [comment]
Confirmed working for me. Thank you!

**Comment by davidgaryesp** [comment]
Neither original (v 1.2.1) nor patch (as posted in this thread) working for me, for latest Obsidian Desktop app (0.15.9) on macOS Catalina.  BUT (1) I am a newbie and (2) something odd going on.  I tried adding Console messages - but they didn't appear.  As an experiment I tried renaming 'main.js' and later even deleted the whole `highlightr-plugin` folder from `myVault/.obsidian/plugins`, restarted Obsidian, but unexpectedly, console messages still indicated `highlightr` was still loading ok.

I don't know if this is normal/odd - but after putting the files back again out of Trash, I uninstalled and reinstalled 'highlightr', but the dates of files in that folder did not change, e.g. it was still the patched version of 'main.js' that was present (nothing was downloaded over the top of them).

It feels (rightly or wrongly) like the plugin folder (or its files) is also getting put into some other plugins location, which the app uses instead.

BTW-FWIW: Chrome released [an update](https://chromereleases.googleblog.com/) - maybe that upset some things?

**Comment by davidgaryesp** [comment]
I found the "other plugins location" issue, put the patched 'main.js' there, and now highlighting works!
The patched version of 'highlightr/main.js' works well for both the desktop app and the iOS app.
_BUT_ - if your vault is in iCloud - the patch needs to be made to _two_ instances of the 'main.js' file - thanks to an iCloud-specific (compatibility?) issue, described below...

Issue: Obsidian and iCloud (where I have put my vault) seem to have disagreements over where things should go, so those things (highlightr code) went to both locations.  This "disagreement" is consistent with what I found yesterday setting up the iOS app to use the same (iCloud-based) vault as the mac desktop.  More detail below...

From the mac's Finder:
The location where if I change the files this has an effect (e.g. patch works):
~/Library/Mobile Documents/iCloud~md~obsidian/Documents/.obsidian/plugins/highlightr-plugin
Presented by Finder as:
<img width="458" alt="image" src="https://user-images.githubusercontent.com/57838694/183300696-bdd93502-66c3-4449-8132-cbe40f572e50.png">

The location where (if I change the files) this has no effect:
~/Library/Mobile Documents/iCloud~md~obsidian/Documents/Obsidian iCloud Vault 001/.obsidian/plugins/highlightr-plugin
Presented by Finder as:
<img width="628" alt="image" src="https://user-images.githubusercontent.com/57838694/183300732-b5d8bc10-b390-4b65-878a-3520c93fa911.png">

Incidentally, as I found when setting up a shared vault for desktop and iOS apps yesterday...
Desktop recognises the first one as "root".
iOS app only recognises the second one.

I guess Obsidian have handled that (in their code) by ensuring plugin downloads get copied to _both_ locations.
In each app/device's case, renaming (only) its particularly-located '.highlightr' files causes it (only) to pop up a "failure to load" message.  Pretty conclusive then.

I imagine the best fix would be for clear and separate concepts of "Obsidian Root in iCloud" and "Obsidian Vault " (within Obsidian Root).  There could be more than one Obsidian Vault in iCloud.  I'm guessing that there might be confusion somewhere(s) "under the hood" about these.

**Comment by heliocarbex** [comment]
> For a workaround for the broken context menu (for the Desktop app), go to your `.obsidian\plugins` folder or `Settings -> Community Plugins -> Installed Plugins` and click the folder icon: ![image](https://user-images.githubusercontent.com/1167333/182479386-cddacec5-9cb2-4ccb-abf6-21b2929f32da.png)
> 
> Go into the `highlightr-plugin` folder and open `main.js`. Search for `menu.showAtPosition` and replace the following:
> 
> ```js
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
> ```
> 
> with
> 
> ```js
>         const orig=menu.__proto__.sort; // Save the original function
>         menu.__proto__.sort = function() {}; // Replace the sort function with a no-op function
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
>        menu.__proto__.sort=orig; // Reset the sort function
> ```
> 
> Basically, the `menu.sort` function relies on menu items which this plugin just ignores. So, when it hits `menu.sort`, the dropdown is rebuilt and the dom (which this plugin manipulates) is reset. This change just makes the `sort` function a no-op.

This worked for me! I installed right now the plugin.

**Comment by Celesica** [comment]
> For a workaround for the broken context menu (for the Desktop app), go to your `.obsidian\plugins` folder or `Settings -> Community Plugins -> Installed Plugins` and click the folder icon: ![image](https://user-images.githubusercontent.com/1167333/182479386-cddacec5-9cb2-4ccb-abf6-21b2929f32da.png)
> 
> Go into the `highlightr-plugin` folder and open `main.js`. Search for `menu.showAtPosition` and replace the following:
> 
> ```js
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
> ```
> 
> with
> 
> ```js
>         const orig=menu.__proto__.sort; // Save the original function
>         menu.__proto__.sort = function() {}; // Replace the sort function with a no-op function
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
>        menu.__proto__.sort=orig; // Reset the sort function
> ```
> 
> Basically, the `menu.sort` function relies on menu items which this plugin just ignores. So, when it hits `menu.sort`, the dropdown is rebuilt and the dom (which this plugin manipulates) is reset. This change just makes the `sort` function a no-op.

Thank you! This fixed mine as well~

**Comment by Paddy1058** [comment]
Yup, works for me too; thanks much!

**Comment by architaktus** [comment]
> For a workaround for the broken context menu (for the Desktop app), go to your `.obsidian\plugins` folder or `Settings -> Community Plugins -> Installed Plugins` and click the folder icon: ![image](https://user-images.githubusercontent.com/1167333/182479386-cddacec5-9cb2-4ccb-abf6-21b2929f32da.png)
> 
> Go into the `highlightr-plugin` folder and open `main.js`. Search for `menu.showAtPosition` and replace the following:
> 
> ```js
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
> ```
> 
> with
> 
> ```js
>         const orig=menu.__proto__.sort; // Save the original function
>         menu.__proto__.sort = function() {}; // Replace the sort function with a no-op function
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
>        menu.__proto__.sort=orig; // Reset the sort function
> ```
> 
> Basically, the `menu.sort` function relies on menu items which this plugin just ignores. So, when it hits `menu.sort`, the dropdown is rebuilt and the dom (which this plugin manipulates) is reset. This change just makes the `sort` function a no-op.

Bravo! It works! Thanks!

**Comment by ikuygithub** [comment]
The last update of this plugin is last December so it is already fossilized.
Not too hopeful

**Comment by timowei** [comment]
> For a workaround for the broken context menu (for the Desktop app), go to your `.obsidian\plugins` folder or `Settings -> Community Plugins -> Installed Plugins` and click the folder icon: ![image](https://user-images.githubusercontent.com/1167333/182479386-cddacec5-9cb2-4ccb-abf6-21b2929f32da.png)
> 
> Go into the `highlightr-plugin` folder and open `main.js`. Search for `menu.showAtPosition` and replace the following:
> 
> ```js
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
> ```
> 
> with
> 
> ```js
>         const orig=menu.__proto__.sort; // Save the original function
>         menu.__proto__.sort = function() {}; // Replace the sort function with a no-op function
>         menu.showAtPosition({
>             x: coords.right + 25,
>             y: coords.top + 20,
>         });
>        menu.__proto__.sort=orig; // Reset the sort function
> ```
> 
> Basically, the `menu.sort` function relies on menu items which this plugin just ignores. So, when it hits `menu.sort`, the dropdown is rebuilt and the dom (which this plugin manipulates) is reset. This change just makes the `sort` function a no-op.

It works on the lastest ver 0.15.9! lovely! Big big hx~~

**Comment by grandgooroo** [comment]
Thx !

**Comment by asoonami** [comment]
The workaround by @remerle fixed the Highlightr issue with context menu not showing on v.0.15.9
Thanks for posting, much appreciated. 

https://github.com/chetachiezikeuzor/Highlightr-Plugin/issues/42#issuecomment-1203250330

**Comment by SimoAbjaou** [comment]
Thanks for the workaround.

**Comment by KraXen72** [comment]
all of these people relying on a workaround that monkey patches an obsidian API function...
@chetachiezikeuzor please merge the pull request to use the new menuItems and this workaround won't be needed
i'd understand if you don't have time to fix the issue, but somebody already made a pull request that fixes it the proper way.
Please do merge it.

Thanks.

**Comment by chetachiezikeuzor** [comment]
Hello friend @gabrielcarus ! Thank you for your notes!

> * Highlightr stopped working in Obsidian v.15.6 (or superior), even the shortcuts.
>   
>   * 7/22/22 - Updated Obsidian to v.15.8, when using the "Ctrl+P" shortcut it works, but the Highlightr pop up don't.

I just released an [update](https://github.com/chetachiezikeuzor/Highlightr-Plugin/releases/tag/1.2.2) for Highlightr that should be much more compatible with Obsidian API. Please let me know if there are any issues with it!
> * An old problem with Highlightr, that I believe you already know, is the space after the marked text. ([Remove extra whitespace #39](https://github.com/chetachiezikeuzor/Highlightr-Plugin/issues/39))

This should be fixed in the update, though I'm still working on a function that is more intelligent.
> * A small detail, that would be nice to analyze, is that cMenu is taking a lot to load.

I'll have to completely redo cMenu at this point. Hoping to get this done soon.
> * Snippets plugin didn't stop, but is having some basic bugs in its structure (not making possible to create new CSS Snippets through the plugin and the "hovering" when using the mouse is also bugged).

I've also released an [update](https://github.com/chetachiezikeuzor/MySnippets-Plugin/releases/tag/1.2.2) for MySnippets. You should be able to create new snippets now.

And thank you everyone for being so patient 🙏🏽!

**Comment by boris73** [comment]
FYI, MacOs 12.5.1, Obsidian 0.15.9 on Highlightr 12.2.2 still has issue with the menu not showing up when you click highlight.

"Ctrl+P" shortcut still works though

**Comment by loune20** [comment]
Sadly [this](https://github.com/chetachiezikeuzor/Highlightr-Plugin/issues/42#issuecomment-1203250330) fix doesn't work for me (the menu doesn't show up and I have the "focus must be on editor" issue), but command palette shortcut works.
This is with Obisidan v0.15.9 and Highlightr 1.2.2. on Ubuntu 20.04.5 LTS

**Comment by HinPeng** [comment]
@boris73 @loune20 
Seems the method above is not working with Highlightr 1.2.2. Downgrade to 1.2.1 and it will work.

**Comment by axelson** [comment]
> FYI, MacOs 12.5.1, Obsidian 0.15.9 on Highlightr 12.2.2 still has issue with the menu not showing up when you click highlight.
> 
> "Ctrl+P" shortcut still works though

That sounds like this new issue: https://github.com/chetachiezikeuzor/Highlightr-Plugin/issues/47

**Comment by loune20** [comment]
> @boris73 @loune20
> Seems the method above is not working with Highlightr 1.2.2. Downgrade to 1.2.1 and it will work.

That worked, thank you so much for pointing that out ! 
I can now highlight and unhighlight normally, except if I select an highlighted word. In that case, I don't have the "_unhighligh_" button in the context menu and if I click the "_highlight_" button, I get the "_focus must be in editor_" error ; I mention it in case it's useful for anyone but that's a great fix for me

**Comment by wlo2** [comment]
Still not working on latest Windows 10 version, please fix!

**Comment by Vanillemilch** [comment]
commenting out _menu.SetUseNativeMenu(false); and restarting worked for me.

![image](https://user-images.githubusercontent.com/73147580/194891552-3861a810-9be6-4841-955c-53bfe041af4a.png)

**Comment by wlo2** [comment]
@Vanillemilch, thank you, works indeed!

**Comment by mnott** [comment]
Oddly, and probably unrelated to the menu, I get yellow as a color, but nothing else - looks as if the tag "mark" has a predefined color and ignores everything else. If I replace it manually by "span", it works. Anyone had that?

**Comment by ichmoimeyo** [comment]
@Vanillemilch, thank you - that fixed it.

**Comment by xinzihou** [comment]
> Sadly [this](https://github.com/chetachiezikeuzor/Highlightr-Plugin/issues/42#issuecomment-1203250330) fix doesn't work for me (the menu doesn't show up and I have the "focus must be on editor" issue), but command palette shortcut works. This is with Obisidan v0.15.9 and Highlightr 1.2.2. on Ubuntu 20.04.5 LTS

Thank you for pointing that out! I have the same issue with Obsidian v0.14.6 (latest version for my device) and Highlightr 1.2.2. Shortcut for opening the highlight options doesn't work, right click menu doesn't work, but shortcuts for individual colours work. I downgraded the plugin to 1.2.1 and it finally works normally. I just upgraded it yesterday and hopefully the bug could be fixed for 1.2.2!

**Comment by xinzihou** [comment]
> 
It worked for me, but weirdly none of the highlighter symbols show colours now.
![image](https://user-images.githubusercontent.com/97024733/236633310-13af207e-a4eb-45ad-82cc-feee9669d5bf.png)
